rust/cyclotron-core/src/janitor_ops.rs

```diff
@@ -0,0 +1,94 @@
+use chrono::{Duration, Utc};
+
+use crate::error::QueueError;
+
+// As a general rule, janitor operations are not queue specific (as in, they don't account
+// for the queue name or waiting_on). We can revisit this later, if we decide we need the
+// ability to do janitor operations on a per-queue basis.
+pub async fn delete_completed_jobs<'c, E>(executor: E) -> Result<u64, QueueError>
+where
+    E: sqlx::Executor<'c, Database = sqlx::Postgres>,
+{
+    let result = sqlx::query!("DELETE FROM cyclotron_jobs WHERE state = 'completed'")
+        .execute(executor)
+        .await
+        .map_err(QueueError::from)?;
+
+    Ok(result.rows_affected())
+}
+
+pub async fn delete_failed_jobs<'c, E>(executor: E) -> Result<u64, QueueError>
+where
+    E: sqlx::Executor<'c, Database = sqlx::Postgres>,
+{
+    let result = sqlx::query!("DELETE FROM cyclotron_jobs WHERE state = 'failed'")
+        .execute(executor)
+        .await
+        .map_err(QueueError::from)?;
+
+    Ok(result.rows_affected())
+}
+
+// Jobs are considered stalled if their lock is held and their last_heartbeat is older than `timeout`.
+// NOTE - because this runs on running jobs, it can stall workers trying to flush updates as it
+// executes. I need to use some of the load generators alongside explain/analyze to optimise this (and
```

Comment:
Just that they mutate on the same rows, and because the janitor ops involve table scans, that can mean a worker task stalls doing the update. I could well be wrong on this though